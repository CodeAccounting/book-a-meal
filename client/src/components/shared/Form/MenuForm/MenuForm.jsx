import React, { Fragment, Component } from 'react';
import moment from 'moment';
import PropTypes from 'prop-types';
import { mealObjPropTypes } from '../../../../helpers/proptypes';
import MiniPreloader from '../../Preloader/MiniPreloader';

/**
 * @exports
 * @extends Component
 * @class MenuForm
 * @returns {JSX} MenuForm
 */
class MenuForm extends Component {
  static propTypes = {
    meals: PropTypes.arrayOf(mealObjPropTypes).isRequired,
    menu: PropTypes.shape({
      date: PropTypes.string,
      meals: PropTypes.arrayOf(mealObjPropTypes),
    }).isRequired,
    isFetching: PropTypes.bool.isRequired,
    fetchMeals: PropTypes.func.isRequired,
    fetchMenu: PropTypes.func.isRequired,
    addMenu: PropTypes.func.isRequired,
    clearMenuError: PropTypes.func.isRequired,
    submitting: PropTypes.bool.isRequired,
    submitError: PropTypes.string
  }

  static defaultProps = {
    submitError: null
  }

  state = {
    date: this.props.menu.date,
    meals: this.props.menu.meals.map(meal => meal.id)
  }

  /**
   * @memberof MenuForm
   * @returns {JSX} MenuForm
   */
  async componentDidMount() {
    this.props.clearMenuError();
    await this.props.fetchMeals();
    await this.props.fetchMenu(this.props.menu.date);
  }

  /**
   * @memberof MenuForm
   * @param {object} event
   * @returns {JSX} MenuForm
   */
  handleChangeDate = (event) => {
    this.props.clearMenuError();
    const date = moment(event.target.value).format('YYYY-MM-DD');

    this.setState({ date }, () => this.props.fetchMenu(date));
  }

  /**
   * @memberof MenuForm
   * @param {object} event
   * @returns {JSX} MenuForm
   */
  handleSelectMeal = (event) => {
    this.props.clearMenuError();
    const { target } = event;
    const { name } = target;

    return target.checked ? this.addMeal(name) : this.removeMeal(name);
  }

  /**
   * @memberof MenuForm
   * @param {object} meal
   * @returns {JSX} MenuForm
   */
  addMeal = (meal) => {
    this.setState(prevState => ({
      meals: [...prevState.meals, meal]
    }));
  }

  /**
   * @memberof MenuForm
   * @param {object} meal
   * @returns {JSX} MenuForm
   */
  removeMeal = (meal) => {
    this.setState(prevState => ({
      meals: prevState.meals.filter(item => item !== meal)
    }));
  }

  /**
   * @memberof MenuForm
   * @param {object} meal
   * @returns {JSX} MenuForm
   */
  submitMenu = () => {
    this.props.addMenu(this.state);
  }

  /**
   * @memberof MenuForm
   * @returns {JSX} MenuForm
   */
  render() {
    const {
      meals, isFetching, submitError, submitting
    } = this.props;

    return (
      <Fragment>
        {(isFetching || submitting) && <div className="text-center"><MiniPreloader /></div>}
        {!isFetching && meals.length === 0 && <p className="text-center">You Do Not Have Any Meals Yet</p>}
        {!isFetching && !submitting && meals.length !== 0 &&
          <Fragment>
            {submitError && <p className="text-center">{submitError}</p>}
            <div className="meal-options-list">
              <div className="form-input">
                <label htmlFor="date">Date</label>
                <input type="date" id="date" name="date" value={this.state.date} onChange={this.handleChangeDate} />
              </div>
              {meals.map(meal => (
                <div className="form-input-checkbox" key={meal.id}>
                  <input
                    type="checkbox"
                    id={meal.id}
                    name={meal.id}
                    checked={this.state.meals.includes(meal.id)}
                    onChange={this.handleSelectMeal}
                  />
                  &nbsp;&nbsp;
                  <label htmlFor="checkbox">{meal.title}</label>
                </div>))}
            </div>
            <button className="btn btn-pri btn-block" onClick={this.submitMenu}>ADD MEAL OPTIONS</button>
          </Fragment>}
      </Fragment>
    );
  }
}

export default MenuForm;
