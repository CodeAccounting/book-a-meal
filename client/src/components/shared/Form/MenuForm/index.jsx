import React, { Fragment, Component } from 'react';
import moment from 'moment';
import PropTypes from 'prop-types';
import { mealObjPropTypes, metadataProps } from '../../../../helpers/proptypes';
import MiniPreloader from '../../Preloader/MiniPreloader';
import InfiniteLoader from '../../InfiniteLoader';

/**
 * @exports
 * @extends Component
 * @class MenuForm
 * @returns {JSX} MenuForm
 */
class MenuForm extends Component {
  static propTypes = {
    meals: PropTypes.arrayOf(mealObjPropTypes).isRequired,
    menu: PropTypes.shape({
      date: PropTypes.string,
      meals: PropTypes.arrayOf(mealObjPropTypes),
    }).isRequired,
    isFetching: PropTypes.bool.isRequired,
    fetchMeals: PropTypes.func.isRequired,
    fetchMenu: PropTypes.func.isRequired,
    addMenu: PropTypes.func.isRequired,
    editMenu: PropTypes.func.isRequired,
    clearMenuError: PropTypes.func.isRequired,
    submitting: PropTypes.bool.isRequired,
    mealsMetadata: metadataProps.isRequired,
    menuMetadata: metadataProps.isRequired,
  }

  /**
   * @static
   * @memberof MenuForm
   * @param {object} props
   * @returns {JSX} MenuForm
   * check if submitError occurs and then map it to the error
   */
  static getDerivedStateFromProps(props) {
    const error = (props.submitError && props.submitError.date) ?
      props.submitError.date.value : props.submitError;

    return { error };
  }

  state = {
    date: this.props.menu.date,
    meals: this.props.menu.meals.map(meal => meal.id),
    error: null,
    formHeight: 0
  }

  /**
   * @memberof MenuForm
   * @returns {JSX} MenuForm
   */
  componentDidMount() {
    this.props.clearMenuError();
    this.props.fetchMeals();
    this.props.fetchMenu();
  }

  /**
   * @memberof MenuForm
   * @returns {func} fetchMeals|fetchMenu
   */
  fetchMoreData = () => {
    if (this.props.mealsMetadata.next) {
      this.props.fetchMeals(this.props.mealsMetadata);
    }

    if (this.props.menuMetadata.next) {
      this.props.fetchMenu(this.props.menu.date, this.props.menuMetadata);
    }

    return null;
  }

  /**
   * @memberof MenuForm
   * @param {object} event
   * @returns {JSX} MenuForm
   */
  clearFormError = () => {
    this.setState({ error: null });
    this.props.clearMenuError();
  }

  /**
   * @memberof MenuForm
   * @param {object} event
   * @returns {JSX} MenuForm
   */
  handleChangeDate = (event) => {
    this.clearFormError();

    const newDate = moment(event.target.value).format('YYYY-MM-DD');
    const currentDate = moment().format('YYYY-MM-DD');
    const date = new Date(currentDate) > new Date(newDate) ? currentDate : newDate;

    this.setState({ date }, () => this.props.fetchMenu(date));
  }

  /**
   * @memberof MenuForm
   * @param {object} event
   * @returns {JSX} MenuForm
   */
  handleSelectMeal = (event) => {
    this.clearFormError();

    const { target } = event;
    const { name } = target;

    return target.checked ? this.addMealToState(name) : this.removeMealFromState(name);
  }

  /**
   * @memberof MenuForm
   * @param {object} meal
   * @returns {JSX} MenuForm
   */
  addMealToState = (meal) => {
    this.setState(prevState => ({
      meals: [...prevState.meals, meal]
    }));
  }

  /**
   * @memberof MenuForm
   * @param {object} meal
   * @returns {JSX} MenuForm
   */
  removeMealFromState = (meal) => {
    this.setState(prevState => ({
      meals: prevState.meals.filter(item => item !== meal)
    }));
  }

  /**
   * @memberof MenuForm
   * @param {object} meal
   * @returns {JSX} MenuForm
   */
  submitMenu = () => {
    const { menu, addMenu, editMenu } = this.props;
    const { meals } = this.state;

    if (meals.length === 0) return this.setState({ error: 'You Cannot Set Empty Meals as a Menu' });

    return !menu.id ? addMenu(this.state) : editMenu(menu.id, { meals: this.state.meals });
  }

  /**
   * @memberof MenuForm
   * @returns {JSX} MenuFormMeals
   */
  renderMeals = () => (this.props.meals.map(meal => (
    <div className="form-input-checkbox" key={meal.id} ref={(elem) => { this.elem = elem; }}>
      <input
        type="checkbox"
        id={meal.id}
        name={meal.id}
        checked={this.state.meals.includes(meal.id)}
        onChange={this.handleSelectMeal}
      />
        &nbsp;&nbsp;
      <label htmlFor="checkbox">{meal.title}</label>
    </div>)))

  /**
   * @memberof MenuForm
   * @returns {JSX} MenuForm
   */
  renderMenuForm = () => (
    <Fragment>
      {this.state.error && <p className="text-center danger">{this.state.error}</p>}
      <div className="meal-options-list">
        <div className="form-input">
          <label htmlFor="date">Date</label>
          <input type="date" id="date" name="date" min={moment().format('YYYY-MM-DD')} value={this.state.date} onChange={this.handleChangeDate} />
        </div>
        <InfiniteLoader
          items={this.renderMeals()}
          metadata={this.props.mealsMetadata}
          height={50 * this.props.meals.length}
          loadMore={this.fetchMoreData}
        />
      </div>
      <button className="btn btn-pri btn-block" onClick={this.submitMenu}>SAVE MEAL OPTIONS</button>
    </Fragment>
  )

  /**
   * @memberof MenuForm
   * @returns {JSX} MenuForm
   */
  render() {
    const { meals, isFetching, submitting } = this.props;

    return (
      <Fragment>
        {(isFetching || submitting) && <div className="text-center"><MiniPreloader /></div>}
        {!isFetching && meals.length === 0 && <p className="text-center">You Do Not Have Any Meals Yet</p>}
        {!isFetching && !submitting && meals.length !== 0 && this.renderMenuForm()}
      </Fragment>
    );
  }
}

export default MenuForm;
