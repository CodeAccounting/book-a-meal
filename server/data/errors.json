{
  "401": "Unauthorized",
  "403": "Forbidden",
  "404": "Not Found",
  "422": "Validation Error",
  "500": "Internal Server Error",
  "empty": "Empty PUT Requests Are Not Allowed"
}