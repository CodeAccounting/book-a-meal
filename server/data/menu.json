[
  {
    "userId": "8356954a-9a42-4616-8079-887a73455a7f",
     "menuId": "a9fa6cb3-9f5e-46fa-b641-388f898ca824",
     "date": "2018-05-06",
     "meals": [
        "81211c24-51c0-46ec-b1e0-18db55880958",
        "36d525d1-efc9-4b75-9999-3e3d8dc64ce3",
        "baa0412a-d167-4d2b-b1d8-404cb8f02631",
        "f9eb7652-125a-4bcb-ad81-02f84901cdc3"
     ],
     "created": "2018-05-06T14:34:90.000",
     "updated": "2018-05-06T14:34:90.000"
  },
  {
    "userId": "8356954a-9a42-4616-8079-887a73455a7f",
     "menuId": "1adfcfe7-c66d-42d2-82fd-39c1decd290a",
     "date": "2018-04-06",
     "meals": [
        "a3c35e8f-da7a-4113-aa01-a9c0fc088539",
        "3ec802c6-0a32-4d29-b27b-42e0f4b532dd"
     ],
     "created": "2018-04-06T14:34:90.000",
     "updated": "2018-04-06T14:34:90.000"
  },
  {
    "userId": "ac1b253c-6b33-439b-ab6f-805a4fdd2e05",
     "menuId": "3a4fb034-6e52-4bde-9abc-91119a39478d",
     "date": "2018-05-07",
     "meals": [
        "72a3417e-45c8-4559-8b74-8b5a61be8614"
     ],
     "created": "2018-05-06T14:34:90.000",
     "updated": "2018-05-06T14:34:90.000"
  },
  {
    "userId": "ac1b253c-6b33-439b-ab6f-805a4fdd2e05",
     "menuId": "f43f3d49-a6c9-476d-b65a-1af772ff0f36",
     "date": "2018-04-06",
     "meals": [
        "72a3417e-45c8-4559-8b74-8b5a61be8614",
        "46ced7aa-eed5-4462-b2c0-153f31589bdd"
     ],
     "created": "2018-05-06T14:34:90.000",
     "updated": "2018-05-06T14:34:90.000"
  },
  {
     "userId": "8356954a-9a42-4616-8079-887a73455a7f",
     "menuId": "15421f7a-0f82-4802-b215-e0e8efb6bfb3",
     "date": "2018-04-27",
     "meals": [
        "91b6e41c-0972-4ac5-86da-4ac1f5226e83",
        "f9eb7652-125a-4bcb-ad81-02f84901cdc3",
        "8a65538d-f862-420e-bcdc-80743df06578"
     ],
     "created": "2018-05-06T14:34:90.000",
     "updated": "2018-05-06T14:34:90.000"
  }
]