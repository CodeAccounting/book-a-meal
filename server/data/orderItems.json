[
  {
     "id": 1,
     "mealId": "46ced7aa-eed5-4462-b2c0-153f31589bdd",
     "orderId": "fb097bde-5959-45ff-8e21-51184fa60c25",
     "quantity": 2,
     "createdAt": "2018-04-06T14:34:90.000",
     "updatedAt": "2018-04-06T14:34:90.000"
  },
  {
     "id": 2,
     "mealId": "36d525d1-efc9-4b75-9999-3e3d8dc64ce3",
     "orderId": "fb097bde-5959-45ff-8e21-51184fa60c25",
     "quantity": 1,
     "createdAt": "2018-04-06T14:34:90.000",
     "updatedAt": "2018-04-06T14:34:90.000"
  }, 
  {
    "id": 3,
    "mealId": "72a3417e-45c8-4559-8b74-8b5a61be8614",
    "orderId": "fb097bde-5959-45ff-8e21-51184fa60c25",
    "quantity": 2,
    "createdAt": "2018-04-06T14:34:90.000",
    "updatedAt": "2018-04-06T14:34:90.000"
  },
  {
    "id": 4,
    "mealId": "a3c35e8f-da7a-4113-aa01-a9c0fc088539",
    "orderId": "6ed0963f-9663-4fe2-8ad4-2f06c6294482",
    "quantity": 2,
    "createdAt": "2018-05-06T14:34:90.000",
    "updatedAt": "2018-05-06T14:34:90.000"
  },
  {
    "id": 5,
    "mealId": "baa0412a-d167-4d2b-b1d8-404cb8f02631",
    "orderId": "6ed0963f-9663-4fe2-8ad4-2f06c6294482",
    "quantity": 2,
    "createdAt": "2018-05-06T14:34:90.000",
    "updatedAt": "2018-05-06T14:34:90.000"
  }
]