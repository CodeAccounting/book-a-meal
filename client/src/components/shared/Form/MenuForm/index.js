import { connect, } from 'react-redux';
import { bindActionCreators } from 'redux';
import MenuForm from './MenuForm';
import { fetchMeals } from '../../../../store/operations/meals';
import { fetchMenu, setCurrentDay, addMenu, editMenu, clearMenuError } from '../../../../store/operations/menu';

const mapStateToProps = state => ({
  isFetching: state.isFetching,
  meals: state.meals.items,
  menu: {
    id: state.menu.id,
    meals: state.menu.meals,
    date: state.menu.currentDay
  },
  submitting: state.menu.working,
  submitError: state.menu.error,
  menuMetadata: state.menu.metadata,
  mealsMetadata: state.meals.metadata
});

const mapDispatchToProps = dispatch => bindActionCreators({
  fetchMenu, fetchMeals, setCurrentDay, addMenu, editMenu, clearMenuError
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(MenuForm);
