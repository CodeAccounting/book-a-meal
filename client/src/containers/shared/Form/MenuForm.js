import { connect, } from 'react-redux';
import { bindActionCreators } from 'redux';
import MenuForm from '../../../components/shared/Form/MenuForm';
import { fetchMeals } from '../../../actions/meals';
import { fetchMenu, setCurrentDay, addMenu, editMenu, clearMenuError } from '../../../actions/menu';

const mapStateToProps = state => ({
  isFetching: state.isFetching,
  meals: state.meals.items,
  menu: {
    id: state.menu.id,
    meals: state.menu.meals,
    date: state.menu.currentDay
  },
  submitting: state.menu.working,
  submitError: state.menu.error,
  menuMetadata: state.menu.metadata,
  mealsMetadata: state.meals.metadata
});

const mapDispatchToProps = dispatch => bindActionCreators({
  fetchMenu, fetchMeals, setCurrentDay, addMenu, editMenu, clearMenuError
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(MenuForm);
