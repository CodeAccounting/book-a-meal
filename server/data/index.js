import users from './users.json';
import meals from './meals.json';
import menu from './menu.json';
import orders from './orders.json';
import errors from './errors.json';

export default {
  users, meals, menu, orders, errors
};
