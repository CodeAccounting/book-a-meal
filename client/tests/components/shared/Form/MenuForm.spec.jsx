import React from 'react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import { initialValues, caterersMealsObj, metadata } from '../../../setup/data';
import MenuForm from '../../../../src/components/shared/Form/MenuForm/MenuForm';
import ConnectedMenuForm from '../../../../src/components/shared/Form/MenuForm/';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const store = mockStore(initialValues);

const props = {
  meals: caterersMealsObj.meals,
  menu: {
    id: '1234',
    date: '2018-06-16',
    meals: caterersMealsObj.meals,
  },
  isFetching: false,
  fetchMeals: jest.fn(),
  fetchMenu: jest.fn(),
  clearMenuError: jest.fn(),
  addMenu: jest.fn(),
  editMenu: jest.fn(),
  submitting: false,
  submitError: null,
  menuMetadata: metadata,
  mealsMetadata: metadata
};

const { now } = Date;

describe('MenuForm', () => {
  beforeAll(() => {
    Date.now = jest.fn(() => 0);
  });

  afterAll(() => {
    jest.clearAllMocks();
    Date.now = now;
  });

  it('renders correctly', () => {
    const wrapper = shallow(<MenuForm {...props} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly when fetching', () => {
    const wrapper = shallow(<MenuForm {...props} isFetching />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly when submitting', () => {
    const wrapper = shallow(<MenuForm {...props} submitting />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly when there is a submit error', () => {
    const wrapper = shallow(<MenuForm {...props} submitError="Error" />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly when there is a date error', () => {
    const errorObj = { date: { value: 'Error' } };
    const wrapper = shallow(<MenuForm {...props} submitError={errorObj} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders correctly when there are no meals', () => {
    const wrapper = shallow(<MenuForm {...props} meals={[]} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('adds meal to state on meal select', () => {
    const wrapper = mount(<MenuForm {...props} />);
    const event = { target: { name: '81211c24-51c0-46ec-b1e0-18db55880958', type: 'checkbox', checked: false } };

    wrapper.find('input').at(1).simulate('change', event);

    const mealAvailInState = wrapper.state().meals.findIndex(meal => meal === '81211c24-51c0-46ec-b1e0-18db55880958');

    expect(mealAvailInState).toEqual(-1);
  });

  it('removes meal from state on meal select', () => {
    const wrapper = mount(<MenuForm {...props} />);
    const falseEvent = { target: { name: '81211c24-51c0-46ec-b1e0-18db55880958', type: 'checkbox', checked: false } };
    const trueEvent = { target: { name: '81211c24-51c0-46ec-b1e0-18db55880958', type: 'checkbox', checked: true } };
    const mealAvailInState = wrapper.state().meals.findIndex(meal => meal === '81211c24-51c0-46ec-b1e0-18db55880958');

    wrapper.find('input').at(1).simulate('change', falseEvent);
    wrapper.find('input').at(1).simulate('change', trueEvent);

    expect(mealAvailInState).toEqual(0);
  });

  it('changes date state on date change', () => {
    // Date.now() =
    const wrapper = mount(<MenuForm {...props} />);
    const event = { target: { value: '2018-04-27' } };

    wrapper.find('input').at(0).simulate('change', event);

    expect(wrapper.state().date).toEqual('2018-04-27');
  });

  it('does not changes date state on date change when date is lower than current date', () => {
    const wrapper = mount(<MenuForm {...props} />);
    const event = { target: { value: '1969-01-01' } };

    wrapper.find('input').at(0).simulate('change', event);

    expect(wrapper.state().date).toEqual('1970-01-01');
  });

  it('submits Menu Form when adding menu', () => {
    const menu = { id: null, date: '2018-06-16', meals: [] };
    const addMenuMock = jest.fn();
    const wrapper = mount(<MenuForm {...props} menu={menu} addMenu={addMenuMock} />);

    wrapper.setState({ meals: caterersMealsObj.meals });

    wrapper.find('button').simulate('click');

    expect(addMenuMock).toHaveBeenCalled();
  });

  it('submits Menu Form when editing', () => {
    const editMenuMock = jest.fn();
    const wrapper = mount(<MenuForm {...props} editMenu={editMenuMock} />);

    wrapper.find('button').simulate('click');

    expect(editMenuMock).toHaveBeenCalled();
  });

  it('does not submit Menu Form when meal state is empty', () => {
    const menu = { id: null, date: '2018-06-16', meals: [] };
    const addMenuMock = jest.fn();
    const wrapper = mount(<MenuForm {...props} menu={menu} addMenu={addMenuMock} />);

    wrapper.find('button').simulate('click');

    expect(addMenuMock).not.toHaveBeenCalled();
    expect(wrapper.find('p.danger')).toBeTruthy();
  });

  it('renders connected component correctly', () => {
    const wrapper = mount(<Provider store={store}><ConnectedMenuForm {...props} /></Provider>);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('calls fetchMeals when fetchMoreData is called and meals metadata has next url', () => {
    const fetchMealsMock = jest.fn();

    const wrapper = mount(<MenuForm
      {...props}
      fetchMeals={fetchMealsMock}
    />);

    wrapper.instance().fetchMoreData();

    expect(fetchMealsMock).toHaveBeenCalledTimes(2);
  });

  it('doesnt call fetchMeals again when fetchMoreData is called and meals metadata does not have next url', () => {
    const fetchMealsMock = jest.fn();

    const wrapper = mount(<MenuForm
      {...props}
      mealsMetadata={{}}
      fetchMeals={fetchMealsMock}
    />);

    wrapper.instance().fetchMoreData();

    expect(fetchMealsMock).toHaveBeenCalledTimes(1);
  });

  it('calls fetchMenu when fetchMoreData is called and menu metadata has next url', () => {
    const fetchMenuMock = jest.fn();

    const wrapper = mount(<MenuForm
      {...props}
      fetchMenu={fetchMenuMock}
    />);

    wrapper.instance().fetchMoreData();

    expect(fetchMenuMock).toHaveBeenCalledTimes(2);
  });

  it('does not call fetchMenu again when fetchMoreData is called and menu metadata does not have next url', () => {
    const fetchMenuMock = jest.fn();

    const wrapper = mount(<MenuForm
      {...props}
      menuMetadata={{}}
      fetchMenu={fetchMenuMock}
    />);

    wrapper.instance().fetchMoreData();

    expect(fetchMenuMock).toHaveBeenCalledTimes(1);
  });
});
