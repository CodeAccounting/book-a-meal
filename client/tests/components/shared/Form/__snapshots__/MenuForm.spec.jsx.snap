// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuForm renders connected component correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(MenuForm)
    addMenu={[MockFunction]}
    clearMenuError={
      [MockFunction] {
        "calls": Array [
          Array [],
          Array [],
          Array [],
          Array [],
          Array [],
          Array [],
          Array [],
          Array [],
          Array [],
          Array [],
          Array [],
          Array [],
          Array [],
          Array [],
          Array [],
          Array [],
        ],
      }
    }
    editMenu={[MockFunction]}
    fetchMeals={
      [MockFunction] {
        "calls": Array [
          Array [],
          Array [],
          Array [],
          Array [],
          Array [],
          Array [],
          Array [],
          Array [],
          Array [],
          Array [],
          Array [],
          Array [],
        ],
      }
    }
    fetchMenu={
      [MockFunction] {
        "calls": Array [
          Array [],
          Array [],
          Array [],
          Array [],
          Array [],
          Array [],
          Array [],
          Array [],
          Array [],
          Array [
            "2018-04-27",
          ],
          Array [],
          Array [],
          Array [],
        ],
      }
    }
    isFetching={false}
    meals={
      Array [
        Object {
          "description": "Jollof Rice, Beef and Plantain. 2 pieces of beef per plate",
          "id": "81211c24-51c0-46ec-b1e0-18db55880958",
          "imageUrl": "http://www.preciouscore.com/wp-content/uploads/2017/11/How-to-cook-jollof-rice-in-the-oven-750x500.jpg",
          "price": "1500.00",
          "title": "Jollof Rice, Beef and Plantain",
          "vegetarian": false,
        },
        Object {
          "description": "Sharwama contains no animal products, perfect for dieters",
          "id": "36d525d1-efc9-4b75-9999-3e3d8dc64ce3",
          "imageUrl": "https://images.pexels.com/photos/461198/pexels-photo-461198.jpeg?auto=compress&cs=tinysrgb&h=350",
          "price": "1200.00",
          "title": "Vegetable Sharwama and Guava Smoothie",
          "vegetarian": true,
        },
      ]
    }
    mealsMetadata={
      Object {
        "next": "next",
        "pages": 2,
        "prev": "prev",
        "total": 5,
      }
    }
    menu={
      Object {
        "date": "2018-06-16",
        "id": "1234",
        "meals": Array [
          Object {
            "description": "Jollof Rice, Beef and Plantain. 2 pieces of beef per plate",
            "id": "81211c24-51c0-46ec-b1e0-18db55880958",
            "imageUrl": "http://www.preciouscore.com/wp-content/uploads/2017/11/How-to-cook-jollof-rice-in-the-oven-750x500.jpg",
            "price": "1500.00",
            "title": "Jollof Rice, Beef and Plantain",
            "vegetarian": false,
          },
          Object {
            "description": "Sharwama contains no animal products, perfect for dieters",
            "id": "36d525d1-efc9-4b75-9999-3e3d8dc64ce3",
            "imageUrl": "https://images.pexels.com/photos/461198/pexels-photo-461198.jpeg?auto=compress&cs=tinysrgb&h=350",
            "price": "1200.00",
            "title": "Vegetable Sharwama and Guava Smoothie",
            "vegetarian": true,
          },
        ],
      }
    }
    menuMetadata={
      Object {
        "next": "next",
        "pages": 2,
        "prev": "prev",
        "total": 5,
      }
    }
    submitError={null}
    submitting={false}
  >
    <MenuForm
      addMenu={[Function]}
      clearMenuError={[Function]}
      editMenu={[Function]}
      fetchMeals={[Function]}
      fetchMenu={[Function]}
      isFetching={false}
      meals={Array []}
      mealsMetadata={Object {}}
      menu={
        Object {
          "date": "1970-01-01",
          "id": null,
          "meals": Array [],
        }
      }
      menuMetadata={Object {}}
      setCurrentDay={[Function]}
      submitError={null}
      submitting={false}
    >
      <p
        className="text-center"
      >
        You Do Not Have Any Meals Yet
      </p>
    </MenuForm>
  </Connect(MenuForm)>
</Provider>
`;

exports[`MenuForm renders correctly 1`] = `
<React.Fragment>
  <React.Fragment>
    <div
      className="meal-options-list"
    >
      <div
        className="form-input"
      >
        <label
          htmlFor="date"
        >
          Date
        </label>
        <input
          id="date"
          min="1970-01-01"
          name="date"
          onChange={[Function]}
          type="date"
          value="2018-06-16"
        />
      </div>
      <InfiniteLoader
        height={100}
        items={
          Array [
            <div
              className="form-input-checkbox"
            >
              <input
                checked={true}
                id="81211c24-51c0-46ec-b1e0-18db55880958"
                name="81211c24-51c0-46ec-b1e0-18db55880958"
                onChange={[Function]}
                type="checkbox"
              />
                
              <label
                htmlFor="checkbox"
              >
                Jollof Rice, Beef and Plantain
              </label>
            </div>,
            <div
              className="form-input-checkbox"
            >
              <input
                checked={true}
                id="36d525d1-efc9-4b75-9999-3e3d8dc64ce3"
                name="36d525d1-efc9-4b75-9999-3e3d8dc64ce3"
                onChange={[Function]}
                type="checkbox"
              />
                
              <label
                htmlFor="checkbox"
              >
                Vegetable Sharwama and Guava Smoothie
              </label>
            </div>,
          ]
        }
        loadMore={[Function]}
        metadata={
          Object {
            "next": "next",
            "pages": 2,
            "prev": "prev",
            "total": 5,
          }
        }
      />
    </div>
    <button
      className="btn btn-pri btn-block"
      onClick={[Function]}
    >
      SAVE MEAL OPTIONS
    </button>
  </React.Fragment>
</React.Fragment>
`;

exports[`MenuForm renders correctly when fetching 1`] = `
<React.Fragment>
  <div
    className="text-center"
  >
    <MiniPreloader />
  </div>
</React.Fragment>
`;

exports[`MenuForm renders correctly when submitting 1`] = `
<React.Fragment>
  <div
    className="text-center"
  >
    <MiniPreloader />
  </div>
</React.Fragment>
`;

exports[`MenuForm renders correctly when there are no meals 1`] = `
<React.Fragment>
  <p
    className="text-center"
  >
    You Do Not Have Any Meals Yet
  </p>
</React.Fragment>
`;

exports[`MenuForm renders correctly when there is a date error 1`] = `
<React.Fragment>
  <React.Fragment>
    <p
      className="text-center danger"
    >
      Error
    </p>
    <div
      className="meal-options-list"
    >
      <div
        className="form-input"
      >
        <label
          htmlFor="date"
        >
          Date
        </label>
        <input
          id="date"
          min="1970-01-01"
          name="date"
          onChange={[Function]}
          type="date"
          value="2018-06-16"
        />
      </div>
      <InfiniteLoader
        height={100}
        items={
          Array [
            <div
              className="form-input-checkbox"
            >
              <input
                checked={true}
                id="81211c24-51c0-46ec-b1e0-18db55880958"
                name="81211c24-51c0-46ec-b1e0-18db55880958"
                onChange={[Function]}
                type="checkbox"
              />
                
              <label
                htmlFor="checkbox"
              >
                Jollof Rice, Beef and Plantain
              </label>
            </div>,
            <div
              className="form-input-checkbox"
            >
              <input
                checked={true}
                id="36d525d1-efc9-4b75-9999-3e3d8dc64ce3"
                name="36d525d1-efc9-4b75-9999-3e3d8dc64ce3"
                onChange={[Function]}
                type="checkbox"
              />
                
              <label
                htmlFor="checkbox"
              >
                Vegetable Sharwama and Guava Smoothie
              </label>
            </div>,
          ]
        }
        loadMore={[Function]}
        metadata={
          Object {
            "next": "next",
            "pages": 2,
            "prev": "prev",
            "total": 5,
          }
        }
      />
    </div>
    <button
      className="btn btn-pri btn-block"
      onClick={[Function]}
    >
      SAVE MEAL OPTIONS
    </button>
  </React.Fragment>
</React.Fragment>
`;

exports[`MenuForm renders correctly when there is a submit error 1`] = `
<React.Fragment>
  <React.Fragment>
    <p
      className="text-center danger"
    >
      Error
    </p>
    <div
      className="meal-options-list"
    >
      <div
        className="form-input"
      >
        <label
          htmlFor="date"
        >
          Date
        </label>
        <input
          id="date"
          min="1970-01-01"
          name="date"
          onChange={[Function]}
          type="date"
          value="2018-06-16"
        />
      </div>
      <InfiniteLoader
        height={100}
        items={
          Array [
            <div
              className="form-input-checkbox"
            >
              <input
                checked={true}
                id="81211c24-51c0-46ec-b1e0-18db55880958"
                name="81211c24-51c0-46ec-b1e0-18db55880958"
                onChange={[Function]}
                type="checkbox"
              />
                
              <label
                htmlFor="checkbox"
              >
                Jollof Rice, Beef and Plantain
              </label>
            </div>,
            <div
              className="form-input-checkbox"
            >
              <input
                checked={true}
                id="36d525d1-efc9-4b75-9999-3e3d8dc64ce3"
                name="36d525d1-efc9-4b75-9999-3e3d8dc64ce3"
                onChange={[Function]}
                type="checkbox"
              />
                
              <label
                htmlFor="checkbox"
              >
                Vegetable Sharwama and Guava Smoothie
              </label>
            </div>,
          ]
        }
        loadMore={[Function]}
        metadata={
          Object {
            "next": "next",
            "pages": 2,
            "prev": "prev",
            "total": 5,
          }
        }
      />
    </div>
    <button
      className="btn btn-pri btn-block"
      onClick={[Function]}
    >
      SAVE MEAL OPTIONS
    </button>
  </React.Fragment>
</React.Fragment>
`;
